### PR TITLE
Fixed dinodex

### PR DESCRIPTION
**Pull request by Zamyatin**

### CODE REVIEW

File: dino_catalog/dinocontroller.rb [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    view_home
+  end
+end
+
+def controller_save_file(location)
+  if @dinodex.dinosaurs.empty?
+    begin
+      raise "You have no data to save!"
+    rescue
+    end
```
Comment (fengb) [inline dino_catalog/dinocontroller.rb:38]: This style of programming is pretty bad. It also does absolutely nothing.

https://en.wikipedia.org/wiki/Coding_by_exception
http://stackoverflow.com/questions/1234343/why-are-empty-catch-blocks-a-bad-idea
http://stackoverflow.com/questions/10594113/bad-idea-to-catch-all-exceptions-in-python

File: dino_catalog/dinoparser.rb [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    @files = files
+    @dinosaurs = []
+  end
+
+  def add(dino)
+    @dinosaurs << Dinosaur.new(standardize(dino))
+  end
+
+  def find_bipeds
+    @dinosaurs.select(&:is_biped?)
```
Comment (fengb) [inline dino_catalog/dinoparser.rb:14]: Don't prepend boolean methods with `is_`. It's a Javaism. The Ruby corollary is just the `?` suffix.

File: dino_catalog/dinoparser.rb [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+  def find_by_period(era)
+    @dinosaurs.select { |dino| dino.period.include? era }
+  end
+
+  def find_by_size!(size)
+    if size.downcase == "big"
+      @dinosaurs.select { |dino| !dino.weight_in_lbs.nil? && dino.weight_in_lbs > 2000 }
+    elsif size.downcase == "small"
+      @dinosaurs.select { |dino| !dino.weight_in_lbs.nil? && dino.weight_in_lbs <= 2000 }
```
Comment (fengb) [inline dino_catalog/dinoparser.rb:29]: This logic is complex and domain specific. Thus, it should be pushed into the model layer:

``` ruby
def small?
  !weight_in_lbs.nil? && weight_in_lbs <= 2000
end
```

File: dino_catalog/dinoparser.rb [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+  end
+
+  def load_dinosaurs
+    @files.each { |file| parse_dinosaurs(file) }
+  end
+
+  def save(filename)
+    unless File.exist?(filename)
+      File.new(filename, "w+")
+    end
```
Comment (fengb) [inline dino_catalog/dinoparser.rb:42]: CSV.open should create the file if it doesn't exist.

File: dino_catalog/dinoparser.rb [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    CSV.open(filename, "w") do |csv|
+      @dinosaurs.each do |dino|
+        csv << [dino.name, dino.period, dino.continent, dino.diet, dino.weight_in_lbs, dino.walking, dino.description]
+      end
+    end
+  end
+
+  def to_json
+    @dinosaurs.each(&:to_json)
```
Comment (fengb) [inline dino_catalog/dinoparser.rb:51]: This doesn't actually create JSON.

File: dino_catalog/dinoparser.rb [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+      continent: args[:continent],
+      diet: args[:diet] || args[:carnivore],
+      weight_in_lbs: args[:weight_in_lbs] || args[:weight],
+      walking: args[:walking],
+      description: args[:description],
+    }
+
+    clean_diet(standard_keys)
+    clean_weight(standard_keys)
+    clean_description(standard_keys)
```
Comment (fengb) [inline dino_catalog/dinoparser.rb:69]: I'd personally use non-mutating helper functions. Example:

``` ruby
return {
  name: args[:name] || args[:genus],
  period: args[:period],
  continent: args[:continent],
  diet: clean_diet(args[:diet] || args[:carnivore]),
  weight_in_lbs: clean_weight(args[:weight_in_lbs] || args[:weight]),
  walking: args[:walking],
  description: clean_description(args[:description]),
}
```

File: dino_catalog/dinosaur.rb [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+  def to_s
+    puts "Name: #{name}" unless @name.nil?
+    puts "Period: #{period}" unless @period.nil?
+    puts "Continent: #{continent}" unless @continent.nil?
+    puts "Diet: #{diet}" unless @diet.nil?
+    puts "Weight: #{weight_in_lbs}" unless @weight_in_lbs.nil?
+    puts "Walking Style: #{walking}" unless @walking.nil?
+    puts "Brief Description: #{description}" unless @description.nil?
+    puts
```
Comment (fengb) [inline dino_catalog/dinosaur.rb:46]: `puts` has no business in the model. This method should return a string.

File: dino_catalog/dinosaur.rb [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    @name = args[:name]
+    @period = args[:period]
+    @continent = args[:continent]
+    @diet = args[:diet]
+    @weight_in_lbs = args[:weight_in_lbs]
+    @walking = args[:walking]
+    @description = args[:description]
+
+    validate_weight!(@weight_in_lbs)
+    validate_description!(@description)
```
Comment (fengb) [inline dino_catalog/dinosaur.rb:14]: Validation only happens in the `initialize` but the fields are writable so this model doesn't really guarantee validness.

File: dino_catalog/runner.rb [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+require 'csv'
+require 'json'
+require_relative 'dinosaur.rb'
+require_relative 'dinoparser.rb'
+require_relative 'dinocontroller.rb'
+require_relative 'dinoviewer.rb'
+
+@messages = ""
+@dinodex = DinoParser.new(['./dinodex.csv', './african_dinosaur_export.csv'])
+@dinodex.load_dinosaurs
```
Comment (fengb) [inline dino_catalog/runner.rb:11]: These variables are globals.

File: dino_catalog/dinoviewer.rb [hunk 10]
```diff
@@ -0,0 +1,90 @@
+def view_home
+  puts @messages if @messages
+  @messages = ""
+  puts "Hello there Dinosaur Hunter! What would you like to do today??"
+  puts "--- Type 'view' to see your Dinosaurs."
+  puts "--- Type 'add' to add a Dinosaur."
+  puts "--- Type 'search' to pull up Dinosaur Search Functions."
+  puts "--- Type 'save' to save your Dinosaurs."
+  puts "--- Type 'exit' to exit the program."
+  input = gets.downcase.chomp
```
Comment (fengb) [inline dino_catalog/dinoviewer.rb:10]: Views are supposed to handle displaying info to the user. Getting input from the user is probably controller logic.

File: dino_catalog/dinoviewer.rb [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+  dino[:weight_in_lbs] = gets.chomp
+  puts "Was your Dinosaur a (Biped) or a (Quadruped)?"
+  dino[:walking] = gets.chomp.capitalize!
+  puts "Write up a little Description if you'd like..."
+  dino[:description] = gets.chomp
+  puts "Here's your Dinosaur:"
+  dino.each { |k, v| puts "#{k}: #{v}" }
+  puts "Would you like to store it in your DinoDex? (Y/n)"
+  response = gets.downcase.chomp
+  controller_add_dino(dino, response)
```
Comment (fengb) [inline dino_catalog/dinoviewer.rb:44]: The view is invoking controller logic, and the controller is invoking view logic. Circular references are usually overly complex and confusing.

File: dino_catalog/dinocontroller.rb [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    when "search"
+      view_search_dinos
+    when "exit"
+      exit
+    else
+      begin
+        raise "That's not a valid option!"
+      rescue
+      end
+      view_home
```
Comment (fengb) [inline dino_catalog/dinocontroller.rb:18]: This is tail recursion at its finest. Ruby also doesn't handle it too well so this is a stack leak.

Comment (Zamyatin) [inline dino_catalog/dinocontroller.rb:18]: AWESOME! I don't know what that means!  I'll add it to the list of things to read about!  :)

**Pull request closed without merging**